needy/caches/directorycache.py

```diff
@@ -0,0 +1,55 @@
+import tarfile
+import os
+from datetime import datetime, timedelta
+
+from ..cache import Cache
+
+
+class DirectoryCache(Cache):
+    def __init__(self, configuration):
+        Cache.__init__(self, configuration)
+
+    @staticmethod
+    def type():
+        return 'directory'
+
+    def description(self):
+        return self.__path
+
+    def archive(self, source, key):
+        destination_path = self.__archive_path(key)
+        parent_path = os.path.dirname(destination_path)
+        if not os.path.exists(parent_path):
+            os.makedirs(parent_path)
+        if os.path.exists(destination_path):
+            os.remove(destination_path)
+        tar = tarfile.open(destination_path, 'w:gz')
+        tar.add(source, arcname='.')
+        tar.close()
+        self.__garbage_collect()
+
+    def unarchive(self, destination, key):
+        tar = tarfile.open(self.__archive_path(key), 'r:gz')
+        tar.extractall(path=destination)
+        tar.close()
+        os.utime(self.__archive_path(key), None)
+
+    def exists(self, key):
+        return os.path.exists(self.__archive_path(key))
+
+    def __garbage_collect(self, min_date=datetime.now()+timedelta(weeks=-2)):
+        for root, directories, filenames in os.walk(self.__archive_path()):
+            for filename in filenames:
+                path = os.path.join(root, filename)
+                try:
+                    mtime = os.path.getmtime(path)
+                except OSError:
+                    mtime = 0
+                last_modified_date = datetime.fromtimestamp(mtime)
+                if last_modified_date < min_date:
+                    logging.debug('Removing older cache item {}'.format(path))
+                    os.remove(path)
```

Comment:
it appears that you're right--there's no official way to extend an object's lifetime.

I don't think the issue of creating an object, having it placed in the cache, and then being deleted by another process is an issue for correctness. Definitely an issue for performance, but I can't think of a way the program breaks as a result of that. Perhaps we're using different definitions of correctness.

Furthermore, the only way that could happen is if a process begins the deletion process and the object to be deleted is deleted and then replaced before the delete command goes through. For long expiration times, it doesn't seem like much of an issue.

That said, if we did support EFS, it appears we would have to use Needy from an EC2 instance or transfer files to an EC2 instance for use on EFS. Unless there's a way to mount EFS remotely, possibly via NFS (didn't look far enough to determine if it was available outside of AWS), I don't know how much adoption that feature would have. In any case, if the only way to utilize it is via a file system mount, this directory cache covers that.

Did you find any way to use EFS easily to satisfy Our needs?